ggml-opencl.c

```diff
@@ -188,35 +189,123 @@ static cl_program build_program_from_source(cl_context ctx, cl_device_id dev, co
+        cl_uint num_devices;
+        clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, NDEV, devices, &num_devices);
+
+        unsigned dev_num;
+        if (sscanf(GGML_OPENCL_DEVICE, " %u", &dev_num) == 1) {
+            if (dev_num >= num_devices) {
+                fprintf(stderr, "ggml_opencl: There is no device %d\n", dev_num);
+                exit(1);
+            } else {
+                device = devices[dev_num];
+                clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL);
+            }
+        } else {
+            for (unsigned i = 0; i < num_devices; i++) {
+                clGetDeviceInfo(devices[i], CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL);
+                if (strstr(text_buffer, GGML_OPENCL_DEVICE) != NULL) {
+                    device = devices[i];
+                    break;
+                }
+            }
+        }
+        if (device == NULL) {
+            fprintf(stderr, "ggml_opencl: no device matching '%s' was found.\n", GGML_OPENCL_DEVICE);
+            exit(1);
+        } else {
+            fprintf(stderr, "ggml_opencl: selecting device: '%s'\n", text_buffer);
+        }
+    }
+
+    cl_context_properties *properties = platform == NULL ? NULL : (cl_context_properties[]){
+        (intptr_t)CL_CONTEXT_PLATFORM, (intptr_t)platform, 0
+    };
+
+    if (device != NULL) {
+        context = clCreateContext(properties, 1, &device, NULL, NULL, &err);
+        CL_CHECK(err, "clCreateContext");
+    } else {
+        context = clCreateContextFromType(properties, CL_DEVICE_TYPE_GPU, NULL, NULL, &err);
+        if (err == CL_DEVICE_NOT_AVAILABLE || err == CL_DEVICE_NOT_FOUND) {
+            context = clCreateContextFromType(properties, CL_DEVICE_TYPE_DEFAULT, NULL, NULL, &err);
+            if (err == CL_DEVICE_NOT_AVAILABLE || err == CL_DEVICE_NOT_FOUND) {
+                context = clCreateContextFromType(properties, CL_DEVICE_TYPE_ALL, NULL, NULL, &err);
+            }
+        }
+        CL_CHECK(err, "clCreateContextFromType");
+    }
+
+    if (device == NULL) {
+        err = clGetContextInfo(context, CL_CONTEXT_DEVICES, sizeof(&device), &device, NULL);
+        CL_CHECK(err, "clGetContextInfo");
+        if (platform == NULL) {
+            err = clGetDeviceInfo(device, CL_DEVICE_PLATFORM, sizeof(&platform), &platform, NULL);
+            CL_CHECK(err, "clGetDeviceInfo");
+        }
+    }
+
+    if (platform != NULL) {
+        clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(text_buffer), &text_buffer, NULL);
+        fprintf(stderr, "ggml_opencl: using platform: '%s'\n", text_buffer);
+    }
+    if (device != NULL) {
+        clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(text_buffer), &text_buffer, NULL);
+        fprintf(stderr, "ggml_opencl: using device: '%s'\n", text_buffer);
+    }
+
+    out_of_order = CL_TRUE;
     queue = clCreateCommandQueue(context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &err);
+    if (err == CL_INVALID_PROPERTY) {
```

Comment:
```
ggml_opencl: selecting platform: 'Apple'
ggml_opencl: selecting device: 'AMD Radeon Pro Vega 20 Compute Engine'
ggml_opencl: using platform: 'Apple'
ggml_opencl: using device: 'AMD Radeon Pro Vega 20 Compute Engine'
ggml_opencl: clCreateCommandQueue error -30 at ggml-opencl.c:307
```
I don't have experience with OpenCL..., should we perhaps also check for `CL_INVALID_VALUE` ?
`if (err == CL_INVALID_PROPERTY || err == CL_INVALID_VALUE)`

With this change, the code then crashes with:
```
GGML_OPENCL_PLATFORM=Apple GGML_OPENCL_DEVICE=AMD ./main 
...
<program source>:3:474: error: automatic variable qualified with an address space
typedef uchar uint8_t; typedef int int32_t; typedef uint uint32_t; struct block_q4_0 { float d; uint8_t qs[16]; }; struct block_q4_1 { float d; float m; uint8_t qs[16]; }; struct __attribute__ ((packed)) block_q5_0 { half d; uint32_t qh; uint8_t qs[16]; }; struct block_q5_1 { half d; half m; uint32_t qh; uint8_t qs[16]; }; struct block_q8_0 { float d; uint8_t qs[16]; }; __kernel void dequantize_row_q4_0(__global struct block_q4_0* x, __global float* y) { constant uint qk = 32; const uint i = get_global_id(0) / qk; const uint j = get_local_id(0); const float d = x[i].d; const int x0 = (x[i].qs[j] & 0xf) - 8; const int x1 = (x[i].qs[j] >> 4) - 8; y[i*qk + j + 0 ] = x0*d; y[i*qk + j + qk/2] = x1*d; } __kernel void dequantize_row_q4_1(__global struct block_q4_1* x, __global float* y) { constant uint qk = 32; const uint i = get_global_id(0) / qk; const uint j = get_local_id(0); const float d = x[i].d; const float m = x[i].m; const int x0 = (x[i].qs[j] & 0xf); const int x1 = (x[i].qs[j] >> 4); y[i*qk + j + 0 ] = x0*d + m; y[i*qk + j + qk/2] = x1*d + m; } __kernel void dequantize_row_q5_0(__global struct block_q5_0* x, __global float* y) { constant uint qk = 32; const uint i = get_global_id(0) / qk; const uint j = get_local_id(0); const float d = vload_half(0, (__global half*) &x[i].d); uint32_t qh = x[i].qh; const uint8_t xh_0 = ((qh >> (j + 0)) << 4) & 0x10; const uint8_t xh_1 = ((qh >> (j + 12)) ) & 0x10; const int32_t x0 = ((x[i].qs[j] & 0xf) | xh_0) - 16; const int32_t x1 = ((x[i].qs[j] >> 4) | xh_1) - 16; y[i*qk + j + 0 ] = x0*d; y[i*qk + j + qk/2] = x1*d; } __kernel void dequantize_row_q5_1(__global struct block_q5_1* x, __global float* y) { constant uint qk = 32; const uint i = get_global_id(0) / qk; const uint j = get_local_id(0); const float d = vload_half(0, (__global half*) &x[i].d); const float m = vload_half(0, (__global half*) &x[i].m); uint32_t qh = x[i].qh; const uint8_t xh_0 = ((qh >> (j + 0)) << 4) & 0x10; const uint8_t xh_1 = ((qh >> (j + 12)) ) & 0x10; const int x0 = (x[i].qs[j] & 0xf) | xh_0; const int x1 = (x[i].qs[j] >> 4) | xh_1; y[i*qk + j + 0 ] = x0*d + m; y[i*qk + j + qk/2] = x1*d + m; } __kernel void dequantize_row_q8_0(__global struct block_q8_0* x, __global float* y) { constant uint qk = 32; const uint i = get_global_id(0) / qk; const uint j = get_local_id(0); const float d = x[i].d; y[i*qk + j] = x[i].qs[j]*d; }
```

Reply:
You are correct.

The error then is about something incompatible in the code, still.

What kind of macs have OpenCL? I have a 2017 macBook at home I could test on? Does it need a GPU or is integrated also usable?

Reply:
It looks like OpenCL should be available on MacBook models >= 2009, see: https://support.apple.com/en-us/HT202823
I'm able to avoid the crash (with a Q4_0 model) by replacing all `constant uint qk = 32;`'s with `const uint qk = 32;`
This "works" for both AMD and Intel UHD, but I'm not seeing any GPU usage in the activity monitor.

Reply:
Yes, I also figured out it's the `constant` that it doesn't like.

I will push fixes shortly.